spock-core/core.gradle

```diff
@@ -15,6 +19,7 @@ configurations { coreConsoleRuntime }
 
 dependencies {
   compile libs.groovy // easiest way to add Groovy dependency to POM
+  compile libs.assertj
```

Comment:
Why did you have to add it? Spock itself shouldn't depend on assertj.

Reply:
I've found the reason. And nice that your verification tool detect it :). AssertJ is a dependency of spock-core's optional dependency `junit-platform-testkit`. Therefor, it should be enough to specify it as an optional dependency in the OSGI manifest.

However, probably accidentally it was [used](https://github.com/spockframework/spock/commit/00417b151cca34665892e040684f90a58f637b9d#diff-5969131316dd5f0d846329f28177f397R3) in the [production](https://github.com/spockframework/spock/commit/00417b151cca34665892e040684f90a58f637b9d#diff-5969131316dd5f0d846329f28177f397R123) code. @Vampire Could you rewrite it to use `StreamSupport`? Otherwise, it could fail in runtime as AssertJ is not available.

Reply:
> I've found the reason. And nice that your verification tool detect it :). AssertJ is a dependency of spock-core's optional dependency junit-platform-testkit. Therefor, it should be enough to specify it as an optional dependency in the OSGI manifest.

I'm glad you found that reason, but there was another! The version of assertj which happened to be in the transitive dependencies was an older version which did not have proper OSGi metadata which meant it did not provide versioned package exports. Adding a newer version allowed for those package imports to be versioned (which is what you ideally always want).

Now, if as you say those should be optional, then it's definitely correct to mark them as such. But if you can just eliminate it that's even better :)

Reply:
Rereading your comment, I added `/org.junit.platform.testkit.*;resolution:=optional/,` as well as `/org.assertj.core.*;resolution:=optional/,`?

the result now is:
```
-runbundles: groovy;version='[2.5.12,2.5.13)',junit-platform-commons;version='[1.7.0,1.7.1)',junit-platform-engine;version='[1.7.0,1.7.1)',junit-platform-launcher;version='[1.7.0,1.7.1)',org.opentest4j;version='[1.2.0,1.2.1)',spock-core;version='[2.0.0,2.0.1)'
```
does that sound correct?

Reply:
> The version of assertj which happened to be in the transitive dependencies was an older version which did not have proper OSGi metadata which meant it did not provide versioned package exports. 

Nice, I didn't know that. The JUnit 5 team bumped the version in master, so it will be fixed in the next version.

> `/org.assertj.core.*;resolution:=optional/,`

What is a convention for playing with transitive optional dependencies (especially with transitive optional dependencies of an optional dependency) in OSGI? Should it be already handled by (in that case) the manifest of `junit5-platform-testkit` (and once upgraded upstream to AssertJ 3.16.2 it could be removed)?

Reply:
I'm the normal case you wouldn't worry about transitive optional packages. The only time is when usage of optional package a requires usage of package b.

Reply:
Ouch, that was non-sense accidental change of course.
Benefit of automatic IDE imports. :-D
I posted PR #1176 that removes that assertj usage, so it should then not be necessary to define it.
But why did you add `org.junit.platform.testkit`? This should also not be used in prod code I think. It is only for testing the Spock engine in Spock tests I think and shouldn't be there in production.

Reply:
It might be good to extract `EmbeddedSpecRunner`-related stuff to a separate module. Let's discuss it on Slack. For the scope of that PR let's keep it as is (but remove the AssertJ from `compile libs.assertj` as it is no longer needed).

@rotty3000 Could you rebase it with master (where AssertJ is no longer needed in production code), preferably applying also a suggestion by Vampire about `slashy-strings`?

Reply:
> the result now is:
>-runbundles: groovy;version='[2.5.12,2.5.13)',junit-platform-commons;version='[1.7.0,1.7.1)',junit-platform-engine;version='[1.7.0,1.7.1)',junit-platform-launcher;version='[1.7.0,1.7.1)',org.opentest4j;version='[1.2.0,1.2.1)',spock-core;version='[2.0.0,2.0.1)'
>does that sound correct?

I don't think so.
There is `junit-platform-launcher` which just is there as dependency of `junit-platform-testkit`.
We don't use it directly except for in a test.
I don't think this should be in Spocks OSGi manifest?

Reply:
I will rebase ASAP

Reply:
assertj is a transitive dependency of `EmbeddedSpecRunner` due to continued use of `import org.junit.platform.testkit.engine.*` Do I leave testkip as a optional dependency for now?

Reply:
Exactly. testkit should for now stay an optional dependency. assertj should be otional automatically by that and thus does not need to be present in the osgi headers explicitly, correct? the move of `EmbeddedSpecRunner` and similar into an own module to remove the optional testkit dependency is handled in a separate issue.

Reply:
fixed